pkg/backend/filestate/backend.go

```diff
@@ -183,12 +189,21 @@ func (b *localBackend) StateDir() string {
 	return workspace.BookkeepingDir
 }
 
+func (b *localBackend) LockDir() string {
+	return filepath.Join(workspace.BookkeepingDir, workspace.LockDir)
+}
+
 func (b *localBackend) ParseStackReference(stackRefName string) (backend.StackReference, error) {
 	return localBackendReference{name: tokens.QName(stackRefName)}, nil
 }
 
 func (b *localBackend) CreateStack(ctx context.Context, stackRef backend.StackReference,
 	opts interface{}) (backend.Stack, error) {
+	err := b.Lock(stackRef)
+	if err != nil {
+		return nil, err
+	}
+	defer b.Unlock(stackRef)
```

Comment:
i'm personally not a fan of this pattern where each method needs to be prefixed with this code.  It feels brittle to me since it would be easy to forget.  I would prefer we do something similar to how we created an intercepting `Bucket` wrapper around go-cloud's abstraction.  In other words, i'd like to see something akin to:

```go
type lockeableBackend struct {
    localBackend *localBackend;
}

// all the methods will then look like this:
func (b *lockeableBackend) CreateStack(ctx context.Context, stackRef backend.StackReference, opts interface{}) (backend.Stack, error) {
    
	err := b.Lock(stackRef)
	if err != nil {
		return nil, err
	}
	defer b.Unlock(stackRef)
        return b.localBackend.CreateStack(ctx, stackRef, opts);
```

This means that we can *easily* see the pattern here and ensure that every operation that needs to be is appropriate locked.  It also keeps the actual localBackend code blissfully unaware of locking, and it means we don't have to constantly be auditing it to make sure some new member does the right thing.

With this new pattern, if we needed to add a new localBackend member and we needed to expose it to someone else, we would have to go through this interface and we would have to ask ourselves "what's the locking rule here".  That's a big plus in my mind.

Reply:
I'm good with this. The `defer b.Unlock` is incompatible with returning an error from `Unlock` as mentioned earlier. Can we forego the error return there in favor of this pattern?